File: frontend/src/plugins/nativeUtils.ts
import { IS_TAURI } from '@/lib/platform'
import { invoke, openDialog, pickDownloadDirectory } from '@/lib/platform-api'

export type DownloadFolderSelectionResponse = {
	uri: string
	path: string
}

export type CopyProgress = {
	totalBytes: string
	progress: number
	cachedPath?: string
}

export class FileSelectedHandler {
	private channelId: string
	private active = true

	constructor(channelId: string) {
		this.channelId = channelId
	}

	public async cancelJob() {
		if (!this.active) return
		await invoke<void>('plugin:native-utils|cancel_job', {
			job: { channelId: this.channelId },
		})
		this.active = false
	}
}

export async function selectDownloadFolder(): Promise<DownloadFolderSelectionResponse | null> {
	if (!IS_TAURI) {
		const path = await pickDownloadDirectory()
		if (!path) return null
		return { uri: path, path }
	}

	return await invoke<DownloadFolderSelectionResponse>(
		'plugin:native-utils|select_download_folder'
	)
}

/** Open the selected Android SAF download folder in a system file manager. */
export async function openDownloadFolder(treeUri: string): Promise<void> {
	if (!IS_TAURI) return

	await invoke<void>('plugin:native-utils|open_download_folder', {
		treeUri,
	})
}

type CopyHandlers = {
	onStart: (path: string, size: bigint) => void
	onEvent: (event: CopyProgress) => void
	onComplete: (path: string) => void
}

function bindCopyChannel(
	channel: { onmessage: (event: CopyProgress) => void },
	handlers: CopyHandlers
) {
	channel.onmessage = (event: CopyProgress) => {
		if (event.progress === 0 && event.cachedPath) {
			handlers.onStart(event.cachedPath, BigInt(event.totalBytes))
		} else if (event.progress === 1 && event.cachedPath) {
			handlers.onComplete(event.cachedPath)
		} else {
			handlers.onEvent(event)
		}
	}
}

export async function selectSendDocument(
	onStart: (path: string, size: bigint) => void,
	onEvent: (event: CopyProgress) => void,
	onComplete: (path: string) => void
): Promise<FileSelectedHandler | null> {
	if (!IS_TAURI) {
		const selected = await openDialog({ multiple: true, directory: false })
		if (!selected) return null
		const paths = Array.isArray(selected) ? selected : [selected]
		for (const path of paths) {
			onStart(path, BigInt(0))
			onComplete(path)
		}
		return null
	}

	const { Channel } = await import('@tauri-apps/api/core')
	const channel = new Channel<CopyProgress>()
	bindCopyChannel(channel, { onStart, onEvent, onComplete })
	const response = await invoke<boolean | undefined>(
		'plugin:native-utils|select_send_document',
		{
			channel: channel,
		}
	)
	if (!response) return null
	return new FileSelectedHandler(String(channel.id))
}

export async function selectSendFolder(
	onStart: (path: string, size: bigint) => void,
	onEvent: (event: CopyProgress) => void,
	onComplete: (path: string) => void
): Promise<FileSelectedHandler | null> {
	if (!IS_TAURI) {
		const selected = await openDialog({ multiple: false, directory: true })
		if (!selected) return null
		const path = Array.isArray(selected) ? selected[0] : selected
		if (!path) return null
		onStart(path, BigInt(0))
		onComplete(path)
		return null
	}

	const { Channel } = await import('@tauri-apps/api/core')
	const channel = new Channel<CopyProgress>()
	bindCopyChannel(channel, { onStart, onEvent, onComplete })
	const response = await invoke<boolean>(
		'plugin:native-utils|select_send_folder',
		{
			channel: channel,
		}
	)
	if (!response) return null
	return new FileSelectedHandler(String(channel.id))
}

/** Consume a pending Android Share-sheet intent (ACTION_SEND). */
export async function consumeShareIntent(
	onStart: (path: string, size: bigint) => void,
	onEvent: (event: CopyProgress) => void,
	onComplete: (path: string) => void
): Promise<FileSelectedHandler | null> {
	if (!IS_TAURI) return null

	const { Channel } = await import('@tauri-apps/api/core')
	const channel = new Channel<CopyProgress>()
	bindCopyChannel(channel, { onStart, onEvent, onComplete })
	const response = await invoke<boolean | undefined>(
		'plugin:native-utils|consume_share_intent',
		{ channel }
	)
	if (!response) return null
	return new FileSelectedHandler(String(channel.id))
}

/** Fired when a share arrives while the app is already open. */
export async function onShareReceived(
	handler: () => void
): Promise<() => void> {
	if (!IS_TAURI) return () => {}

	const { addPluginListener } = await import('@tauri-apps/api/core')
	const listener = await addPluginListener(
		'native-utils',
		'shareReceived',
		() => {
			handler()
		}
	)
	return () => {
		void listener.unregister()
	}
}

File: src-tauri/plugins/tauri-plugin-native-utils/android/src/main/java/com/altsendme/plugin/native_utils/NativeUtils.kt
package com.altsendme.plugin.native_utils

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.provider.DocumentsContract
import android.webkit.WebView
import androidx.activity.result.ActivityResult
import androidx.annotation.Keep
import app.tauri.annotation.ActivityCallback
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Channel
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.cancelAndJoin
import kotlinx.coroutines.launch
import java.io.File
import java.io.IOException
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicReference

@InvokeArg
class SelectorArgs {
    lateinit var channel: Channel
}

@InvokeArg
class CancelJobArgs(
    var channelId: Long = 0
)

@InvokeArg
class ExportToTreeArgs {
    var treeUri: String = ""
    var sourceDir: String = ""
}

@InvokeArg
class OpenDownloadFolderArgs {
    var treeUri: String = ""
}

@Keep
data class DownloadFolderSelectionResponse(
    val uri: String,
    val path: String,
)

@TauriPlugin
class NativeUtils(private val activity: Activity) : Plugin(activity) {
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private val jobs = ConcurrentHashMap<Long, Pair<Job, String>>()
    private val pendingShareUri = AtomicReference<Uri?>(null)

    companion object {
        private const val RW_PERMISSION_FLAGS =
            Intent.FLAG_GRANT_WRITE_URI_PERMISSION or Intent.FLAG_GRANT_READ_URI_PERMISSION
        private const val SHARE_RECEIVED_EVENT = "shareReceived"
    }

    private val consumedShareUris =
        java.util.Collections.synchronizedSet(mutableSetOf<String>())

    @Command
    fun select_download_folder(invoke: Invoke) = startActivityForResult(
        invoke,
        Intent(Intent.ACTION_OPEN_DOCUMENT_TREE),
        this::handleDownloadFolderSelection.name
    )

    @Command
    fun select_send_document(invoke: Invoke) = startActivityForResult(
        invoke,
        Intent(Intent.ACTION_OPEN_DOCUMENT).apply {
            type = "*/*"
        },
        this::handleSendSelection.name
    )

    @Command
    fun select_send_folder(invoke: Invoke) = startActivityForResult(
        invoke,
        Intent(Intent.ACTION_OPEN_DOCUMENT_TREE),
        this::handleSendSelection.name
    )

    @Command
    fun consume_share_intent(invoke: Invoke) {
        val args = invoke.parseArgs(SelectorArgs::class.java)
        val uri = takePendingOrIntentShare()
            ?: return invoke.resolveObject(false)

        startUriCopy(uri, args.channel)
        invoke.resolveObject(true)
    }

    @Command
    fun cancel_job(invoke: Invoke) {
        val args = invoke.parseArgs(CancelJobArgs::class.java)
        val channelId = args.channelId
        val (job, tempFolder) = jobs.remove(channelId)
            ?: return invoke.reject("Trying to cancel a non existing job")
        scope.launch {
            try {
                job.cancelAndJoin()
                File(tempFolder).deleteRecursively()
                invoke.resolve()
            } catch (e: Exception) {
                invoke.reject(e.message)
            }
        }
    }

    @Command
    fun export_to_tree(invoke: Invoke) {
        val args = invoke.parseArgs(ExportToTreeArgs::class.java)
        scope.launch {
            try {
                val treeUri = Uri.parse(args.treeUri)
                val sourceDir = File(args.sourceDir)
                val result = exportDirectoryToTree(activity, treeUri, sourceDir)
                invoke.resolveObject(result)
            } catch (e: SecurityException) {
                invoke.reject(e.message ?: "SAF permission denied")
            } catch (e: Exception) {
                invoke.reject(e.message ?: "Failed to export to selected folder")
            }
        }
    }

    @Command
    fun open_download_folder(invoke: Invoke) {
        val args = invoke.parseArgs(OpenDownloadFolderArgs::class.java)
        val treeUriString = args.treeUri.trim()
        if (treeUriString.isEmpty()) {
            return invoke.reject("No download folder URI available")
        }

        try {
            val treeUri = Uri.parse(treeUriString)
            if (!DocumentsContract.isTreeUri(treeUri)) {
                return invoke.reject("Invalid download folder URI")
            }

            val docId = DocumentsContract.getTreeDocumentId(treeUri)
            val documentUri = DocumentsContract.buildDocumentUriUsingTree(treeUri, docId)
            val intent = Intent(Intent.ACTION_VIEW).apply {
                setDataAndType(documentUri, DocumentsContract.Document.MIME_TYPE_DIR)
                addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }

            try {
                activity.startActivity(intent)
            } catch (_: android.content.ActivityNotFoundException) {
                activity.startActivity(Intent.createChooser(intent, null))
            }
            invoke.resolve()
        } catch (e: Exception) {
            invoke.reject(e.message ?: "Failed to open download folder")
        }
    }

    @ActivityCallback
    fun handleDownloadFolderSelection(invoke: Invoke, result: ActivityResult) {
        if (Activity.RESULT_OK != result.resultCode) return invoke.resolve(null)

        val uri = result.data?.data ?: return invoke.resolve(null)

        try {
            activity.contentResolver.takePersistableUriPermission(uri, RW_PERMISSION_FLAGS)

            invoke.resolveObject(
                DownloadFolderSelectionResponse(
                    uri.toString(),
                    uri.extractFolderOsPath(),
                )
            )

            activity.contentResolver.persistedUriPermissions.stream()
                .filter { it.uri != uri }
                .forEach {
                    activity.contentResolver.releasePersistableUriPermission(
                        it.uri,
                        RW_PERMISSION_FLAGS
                    )
                }
        } catch (e: Exception) {
            invoke.reject(e.message)
        }
    }

    @ActivityCallback
    fun handleSendSelection(invoke: Invoke, result: ActivityResult) {
        val args = invoke.parseArgs(SelectorArgs::class.java)
        val channel = args.channel

        if (Activity.RESULT_OK != result.resultCode) return invoke.resolveObject(false)

        val uri = result.data?.data ?: return invoke.resolveObject(false)

        startUriCopy(uri, channel)
        invoke.resolveObject(true)
    }

    override fun load(webView: WebView) {
        super.load(webView)

        // Cold start: capture share URI before / as the frontend mounts.
        // Skip wiping file_cache when a share is pending so cleanup cannot race the copy.
        val shareUri = peekShareUri(activity.intent)
        if (shareUri != null) {
            pendingShareUri.set(shareUri)
            // Notify after the WebView can register plugin listeners (cold-start race).
            webView.post {
                trigger(SHARE_RECEIVED_EVENT, JSObject())
            }
        } else {
            scope.launch {
                activity.cacheDir.resolve("file_cache").deleteRecursively()
            }
        }
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        // Without this, activity.intent stays the old MAIN launcher intent under singleTask.
        activity.intent = intent

        val uri = peekShareUri(intent) ?: return
        pendingShareUri.set(uri)
        trigger(SHARE_RECEIVED_EVENT, JSObject())
    }

    override fun onResume() {
        super.onResume()
        // Safety net: if the frontend missed the first event (listener not ready yet),
        // re-advertise any still-unconsumed share when we come to the foreground.
        val uri = peekShareUri(activity.intent) ?: return
        pendingShareUri.compareAndSet(null, uri)
        if (pendingShareUri.get() != null) {
            trigger(SHARE_RECEIVED_EVENT, JSObject())
        }
    }

    override fun onDestroy() {
        jobs.forEach { _, (job, tempFolder) ->
            try {
                job.cancel()
                File(tempFolder).deleteRecursively()
            } catch (_: Exception) {
            }
        }

        scope.cancel()
        super.onDestroy()
    }

    private fun startUriCopy(uri: Uri, channel: Channel) {
        val path = listOf(
            activity.cacheDir.absolutePath,
            "file_cache",
            System.currentTimeMillis().toString(),
        ).joinToString(File.separator)

        val tempFolder = File(path)

        val job = scope.launch(start = CoroutineStart.LAZY) {
            try {
                tempFolder.parentFile?.mkdirs()
                    ?: throw IOException("Unable to create parent folders for ${tempFolder.absolutePath}")

                copyUri(activity, uri, tempFolder).collect {
                    channel.send(it.toJSObject())
                }
            } catch (e: Exception) {
                tempFolder.deleteRecursively()
                channel.send(
                    JSObject().apply {
                        put("error", e.message ?: "Failed to copy shared file")
                        put("progress", -1.0)
                        put("copiedBytes", "0")
                        put("totalBytes", "0")
                    }
                )
            } finally {
                jobs.remove(channel.id)
            }
        }

        jobs[channel.id] = job to tempFolder.absolutePath
        job.start()
    }

    @Synchronized
    private fun takePendingOrIntentShare(): Uri? {
        pendingShareUri.getAndSet(null)?.let { uri ->
            markShareConsumed(uri)
            return uri
        }
        return takeAndMarkShareUri(activity.intent)
    }

    private fun peekShareUri(intent: Intent?): Uri? {
        if (intent == null || intent.action != Intent.ACTION_SEND) {
            return null
        }
        val uri = extractShareUri(intent) ?: return null
        if (consumedShareUris.contains(uri.toString())) {
            return null
        }
        return uri
    }

    private fun takeAndMarkShareUri(intent: Intent?): Uri? {
        val uri = peekShareUri(intent) ?: return null
        markShareConsumed(uri)
        return uri
    }

    private fun markShareConsumed(uri: Uri) {
        consumedShareUris.add(uri.toString())
    }

    private fun extractShareUri(intent: Intent): Uri? {
        parcelableStreamExtra(intent)?.let { return it }

        when (val stream = intent.extras?.get(Intent.EXTRA_STREAM)) {
            is Uri -> return stream
            is String -> if (stream.isNotBlank()) return Uri.parse(stream)
        }

        val clip = intent.clipData
        if (clip != null && clip.itemCount > 0) {
            clip.getItemAt(0)?.uri?.let { return it }
        }

        intent.data?.let { return it }

        return null
    }

    @Suppress("DEPRECATION")
    private fun parcelableStreamExtra(intent: Intent): Uri? {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            intent.getParcelableExtra(Intent.EXTRA_STREAM, Uri::class.java)
        } else {
            intent.getParcelableExtra(Intent.EXTRA_STREAM) as? Uri
        }
    }
}

fun Uri.extractFolderOsPath(): String {
    require(DocumentsContract.isTreeUri(this))

    val path = this.path
        ?: throw IOException("Unable to get path from selected download folder uri: $this")
    val baseExternalPath = Environment.getExternalStorageDirectory().path
    return try {
        val docId = DocumentsContract.getTreeDocumentId(this)
        val segments = docId.split(":")
        when {
            "primary" == segments[0] && segments.size > 1 -> "${baseExternalPath}/${segments[1]}"
            "primary" == segments[0] -> baseExternalPath
            segments.size > 1 -> "/storage/${segments[0]}/${segments[1]}"
            else -> "/storage/${segments[0]}/"
        }
    } catch (_: Exception) {
        path
    }
}
